Tests/Test_Renode_MotorSimulator.c

```diff
@@ -0,0 +1,163 @@
+#include "os.h"
+#include "CANbus.h"
+#include <stdio.h>
+#include "Tasks.h"
+#include <math.h>
+
+// Global variables
+static float setpointVelocity = 0.0f; 
+static float velocity = 0.0f;
+static float current = 0.0f; 
+static float total_accel = 0.0f;
+static CANDATA_t oldCD;
+static uint8_t cycle_ctr = 0;
+static uint16_t motor_force = 0;
+static uint8_t forceLUTIndex = 0;
+
+// FORCE LUT 
+// FORCELUT Size
+#define FORCELUT_SIZE 21 
+// Force Lookup Table: Indexed by Current%toIndex, returns a force applied by current % (in Newtons)
+static const int FORCELUT[FORCELUT_SIZE] = {
+    0,
+    309,
+    618,
+    926,
+    1235,
+    1544,
+    1853,
+    2162,
+    2471,
+    2779,
+    3088,
+    3397,
+    3706,
+    4015,
+    4324,
+    4632,
+    4941,
+    5250,
+    5559,
+    5868,
+    6176
+};
+
+// Macros for calculating the velocity of the car
+#define MS_TIME_DELAY_MS 100
+#define DECELERATION 2.0 // In m/s^2
+#define CAR_MASS_KG 270
+#define MAX_VELOCITY 20000.0f // rpm
+
+//linear acceleration used for velocity control mode
+static int linear_accel = FORCELUT[20]/(CAR_MASS_KG * 5);
+
+static OS_TCB Task1TCB;
+static CPU_STK Task1Stk[DEFAULT_STACK_SIZE];
+
+// Helper function to convert from current to desired FORCELUT index
+inline int currentPercentToIndex(float currentPercent){
+    return ((int)(currentPercent*(FORCELUT_SIZE - 1)));
+}    
+
+void Task1(void *arg)
+{
+    while (1)
+    {
+        OS_ERR err;
+        CANDATA_t newCD;
+
+        ErrorStatus error = CANbus_Read(&newCD, CAN_NON_BLOCKING, MOTORCAN); // returns data value into newCD
+        if (error == ERROR) // If there is no new value to read, use the old CAN data
+        {
+            newCD = oldCD;
+        }
+
+        // Parsing Measurements
+        memcpy(&setpointVelocity, &newCD.data[0], sizeof setpointVelocity); // in RPM 
+        memcpy(&current, &newCD.data[4], sizeof current); // Percent from 0.0 to 1.0
+        
+        if (abs(setpointVelocity) == MAX_VELOCITY){
+
+            forceLUTIndex = currentPercentToIndex(current); // Converts current to index for FORCELUT
+            motor_force = FORCELUT[forceLUTIndex];
+
+            // Net acceleration is dependent on the force from the motor (based on current), mass of the car, 
+            // and resistive forces which are being substituted with a constant 2m/s^2 negative acceleration
+            if (setpointVelocity < 0) {
+                total_accel = ((float)motor_force / CAR_MASS_KG - DECELERATION);
```

Comment:
`total_accel` might be negative when `motor_force` is low, which probably isn't desirable. Since the deceleration force is being modeled as a constant, it's probably best to clamp `total_accel` to be at least zero.

Reply:
how exactly would one go about doing so?

Reply:
Roie is saying that total_accel shouldn't be negative, so basically if it is negative just set the value of total_accel to 0. However, don't we need to allow for reverse drive? @Cam0Cow

Reply:
They have a separate case for that. I think I've phrased it poorly, what I meant is that total_accel shouldn't be large enough to make the car go in reverse if it's currently going forward. total_accel will be negative when the car is just coasting.